app/controllers/incomeSources/add/AddIncomeSourceStartDateController.scala

```diff
@@ -121,34 +120,34 @@ class AddIncomeSourceStartDateController @Inject()(authenticate: AuthenticationP
 
     val messagesPrefix = incomeSourceType.startDateMessagesPrefix
 
-    Future.successful(
       if (isEnabled(IncomeSources)) {
         form(messagesPrefix).bindFromRequest().fold(
           formWithErrors =>
-            BadRequest(
+            Future.successful(BadRequest(
               addIncomeSourceStartDate(
                 isAgent = isAgent,
                 form = formWithErrors,
                 backUrl = getBackUrl(incomeSourceType, isAgent, isUpdate),
                 postAction = getPostAction(incomeSourceType, isAgent, isUpdate),
                 messagesPrefix = messagesPrefix
               )
-            ),
-          formData =>
-            Redirect(getSuccessUrl(incomeSourceType, isAgent, isUpdate))
-              .addingToSession(
-                (incomeSourceType match {
-                  case SelfEmployment =>
-                    SessionKeys.addBusinessStartDate
-                  case UkProperty =>
-                    SessionKeys.addUkPropertyStartDate
-                  case ForeignProperty =>
-                    SessionKeys.foreignPropertyStartDate
-                }) -> formData.date.toString
-              )
+            )),
+          formData => {
+            val key = incomeSourceType match {
+              case SelfEmployment =>
+                SessionKeys.addBusinessStartDate
+              case UkProperty =>
+                SessionKeys.addUkPropertyStartDate
+              case ForeignProperty =>
+                SessionKeys.foreignPropertyStartDate
+            }
+            sessionService.set(key, formData.date.toString, Redirect(getSuccessUrl(incomeSourceType, isAgent, isUpdate))) map {
```

Comment:
if you can move any call to sessionService into a separate method within a controller